pyqtorch/circuit.py

```diff
@@ -215,6 +215,114 @@ def sample(probs: Tensor) -> Counter:
             return list(map(lambda p: sample(p), probs))
 
 
+class DropoutQuantumCircuit(QuantumCircuit):
+    """Creates a quantum circuit able to perform quantum dropout, based on the work of https://arxiv.org/abs/2310.04120.
+    Args:
+        dropout_mode (DropoutMode): type of dropout to perform. Defaults to DropoutMode.ROTATIONAL
+        dropout_prob (float): dropout probability. Defaults to 0.06.
+    """
+
+    def __init__(
+        self,
+        n_qubits: int,
+        operations: list[Module],
+        dropout_mode: DropoutMode = DropoutMode.ROTATIONAL,
+        dropout_prob: float = 0.06,
+    ):
+        super().__init__(n_qubits, operations)
+        self.dropout_mode = dropout_mode
+        self.dropout_prob = dropout_prob
+
+        self.dropout_fn = getattr(self, dropout_mode)
+
+    def forward(
+        self, state: State, values: dict[str, Tensor] | ParameterDict = {}
+    ) -> State:
+        if self.training:
+            state = self.dropout_fn(state, values)
+        else:
+            for op in self.operations:
+                state = op(state, values)
+        return state
+
+    def rotational_dropout(
+        self, state: State = None, values: dict[str, Tensor] | ParameterDict = {}
+    ) -> State:
+        """Randomly drops entangling rotational gates.
+
+        Args:
+            state (State, optional): pure state vector . Defaults to None.
+            values (dict[str, Tensor] | ParameterDict, optional): gate parameters. Defaults to {}.
+
+        Returns:
+            State: pure state vector
+        """
+        for op in self.operations:
+            if (
+                (hasattr(op, "param_name"))
+                and (values[op.param_name].requires_grad)
+                and not (int(1 - bernoulli(tensor(self.dropout_prob))))
+            ):
+                continue
+
+            else:
+                state = op(state, values)
```

Comment:
I think I would simplify it by doing something like:

```suggestion
            has_param_name = hasattr(op, "param_name")
            requires_grad = values[op.param_name].requires_grad
            not_<name> = not (int(1 - bernoulli(tensor(self.dropout_prob)))
            if not (has_param_name and requires_grad not_<name>):
                state = op(state, values)
```

Reply:
unfortunately I don't believe we can predicates beforehand since "has_param_name = hasattr(op, "param_name")" will return an error for gates which do not have the param_name attribute such as the CNOT gate. The if statement avoids this issue since the statements are evaluated sequentially. However, I can re-write this so we no longer have the continue statement and just one if statement.